File: code/ch07/go/test_server_stub.go
// §7.5.2 Deploying Test Servers — a stub HTTP server programmed to return 500,
// asserting the clinic client's fallback behavior.
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
)

func alwaysFail(w http.ResponseWriter, _ *http.Request) { http.Error(w, "boom", 500) }
func fetchAppointments(baseURL string) []byte {
	resp, err := http.Get(baseURL + "/appointments")
	if err != nil || resp.StatusCode >= 400 {
		return nil // fallback: empty schedule, not a crash
	}
	body, _ := io.ReadAll(resp.Body)
	return body
}

func main() {
	stub := httptest.NewServer(http.HandlerFunc(alwaysFail))
	defer stub.Close()
	if len(fetchAppointments(stub.URL)) != 0 {
		panic("client should fall back to an empty schedule on 500")
	}
	fmt.Println("client fell back to an empty schedule on 500")
}

File: code/ch07/python/test_server_stub.py
# §7.5.2 Deploying Test Servers — a stub HTTP server programmed to return 500,
# asserting the clinic client's fallback behavior.
import http.server, threading, urllib.error, urllib.request

class AlwaysFail(http.server.BaseHTTPRequestHandler):
  def do_GET(self):
    self.send_error(500)
  def log_message(self, *_):           # keep test output clean
    pass

def fetch_appointments(base_url):
  try:
    with urllib.request.urlopen(base_url + "/appointments") as resp:
      return resp.read()
  except urllib.error.HTTPError:
    return []                        # fallback: empty schedule, not a crash

stub = http.server.HTTPServer(("127.0.0.1", 0), AlwaysFail)
threading.Thread(target=stub.serve_forever, daemon=True).start()

assert fetch_appointments(f"http://127.0.0.1:{stub.server_port}") == []
stub.shutdown()
print("client fell back to an empty schedule on 500")
